### ISSUE TITLE
UTS Machine Learning

### ISSUE THREAD

**Issue by divadwisectiosetiawan-arc** [divadwisectiosetiawan-arc/iot-phi#1]
`# -*- coding: utf-8 -*-
"""UTS Machine Learning.ipynb
```

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/1Bv-qJIOBSCDwV55g3_i_YGRI3h0rzdXQ
"""

```
!pip install imbalanced-learn -q

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

from sklearn.preprocessing import LabelEncoder, OrdinalEncoder, MinMaxScaler
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import (confusion_matrix, classification_report,
                             accuracy_score, ConfusionMatrixDisplay)
from imblearn.over_sampling import SMOTE

print('Semua library berhasil diimport!')

df = pd.read_csv('/content/Social_media_impact_on_life.csv')
print(f'Shape dataset: {df.shape}')
df.head()

print('=== Info Dataset ===')
df.info()
print()
print('=== Statistik Deskriptif ===')
df.describe()

print('=== Nilai Unik Tiap Kolom Kategorikal ===')
cat_cols = ['Gender', 'Academic_Level', 'Most_Used_Platform',
            'Affects_Academic_Performance', 'Overall_Impact']
for col in cat_cols:
    print(f'\n[{col}] — {df[col].nunique()} nilai unik:')
    print(df[col].value_counts().to_string())

class_dist = df['Overall_Impact'].value_counts()
print('=== Distribusi Kelas (Sebelum Penyeimbangan) ===')
print(class_dist)
print(f'\nRasio kelas: {class_dist.values}')

fig, axes = plt.subplots(1, 2, figsize=(12, 4))

# Bar plot
colors = ['#E24B4A', '#888780', '#1D9E75']
axes[0].bar(class_dist.index, class_dist.values, color=colors, edgecolor='white', linewidth=1.5)
axes[0].set_title('Distribusi Kelas Overall_Impact', fontsize=13, fontweight='bold')
axes[0].set_xlabel('Kelas')
axes[0].set_ylabel('Jumlah Data')
for i, (label, val) in enumerate(zip(class_dist.index, class_dist.values)):
    axes[0].text(i, val + 10, str(val), ha='center', fontweight='bold')

# Pie chart
axes[1].pie(class_dist.values, labels=class_dist.index, autopct='%1.1f%%',
            colors=colors, startangle=90, textprops={'fontsize': 11})
axes[1].set_title('Proporsi Kelas Overall_Impact', fontsize=13, fontweight='bold')

plt.suptitle('⚠️ Data Tidak Seimbang: Negative (55%) >> Neutral (15%)',
             fontsize=12, color='#E24B4A', y=1.02)
plt.tight_layout()
plt.show()

print('\n Kesimpulan: Data TIDAK seimbang. Kelas Negative sangat dominan.')
print('   Akan digunakan metode SMOTE (Synthetic Minority Over-sampling Technique)')
print('   untuk menyeimbangkan jumlah kelas.')

print('=== Pengecekan Missing Values ===')
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
missing_df = pd.DataFrame({'Jumlah Missing': missing, 'Persentase (%)': missing_pct})
print(missing_df)
print(f'\nTotal missing values: {missing.sum()}')

if missing.sum() == 0:
    print('\n✅ Tidak ada missing values. Dataset bersih!')
else:
    print('\n⚠️ Terdapat missing values. Mengisi dengan:')
    print('   - Kolom numerik: median')
    print('   - Kolom kategorikal: modus')
    for col in df.columns:
        if df[col].isnull().sum() > 0:
            if df[col].dtype in ['float64', 'int64']:
                df[col].fillna(df[col].median(), inplace=True)
            else:
                df[col].fillna(df[col].mode()[0], inplace=True)
    print('   Selesai mengisi missing values!')

df_enc = df.drop(columns=['Student_ID']).copy()  # Hapus ID, tidak relevan

# 1. Binary Encoding
df_enc['Gender'] = df_enc['Gender'].map({'Male': 0, 'Female': 1})
df_enc['Affects_Academic_Performance'] = df_enc['Affects_Academic_Performance'].map({'No': 0, 'Yes': 1})

# 2. Ordinal Encoding — Academic_Level (ada urutan)
ordinal_enc = OrdinalEncoder(categories=[['High School', 'Undergraduate', 'Graduate']])
df_enc['Academic_Level'] = ordinal_enc.fit_transform(df_enc[['Academic_Level']]).astype(int)

# 3. Label Encoding — Most_Used_Platform (nominal)
le_platform = LabelEncoder()
df_enc['Most_Used_Platform'] = le_platform.fit_transform(df_enc['Most_Used_Platform'])
platform_map = dict(zip(le_platform.classes_, le_platform.transform(le_platform.classes_)))

# 4. Frequency Encoding — Country (111 kategori unik)
country_freq = df['Country'].value_counts(normalize=True)
df_enc['Country'] = df_enc['Country'].map(country_freq)

# 5. Ordinal Encoding — Label/Target
label_map = {'Negative': 0, 'Neutral': 1, 'Positive': 2}
df_enc['Overall_Impact'] = df_enc['Overall_Impact'].map(label_map)

print('=== Ringkasan Encoding ===')
print('Binary Encoding  : Gender, Affects_Academic_Performance')
print('Ordinal Encoding : Academic_Level (High School=0, Undergraduate=1, Graduate=2)')
print('Label Encoding   : Most_Used_Platform')
print('Frequency Enc.   : Country (proporsi frekuensi)')
print('Target Mapping   : Negative=0, Neutral=1, Positive=2')
print(f'\nMapping Platform: {platform_map}')
print()
print('=== Preview Data Setelah Encoding ===')
df_enc.head()

X = df_enc.drop(columns=['Overall_Impact'])
y = df_enc['Overall_Impact']

scaler = MinMaxScaler()
X_scaled = pd.DataFrame(scaler.fit_transform(X), columns=X.columns)

print('=== Data Setelah Normalisasi Min-Max Scaling ===')
print('Semua nilai berada pada rentang [0, 1]')
print()
print(X_scaled.describe().round(4))

# Gunakan fitur numerik kontinu saja untuk pairplot agar lebih readable
df_plot = df[['Age', 'Avg_Daily_Usage_Hours', 'Sleep_Hours_Per_Night',
              'Mental_Health_Score', 'Overall_Impact']].copy()

palette = {'Negative': '#E24B4A', 'Neutral': '#888780', 'Positive': '#1D9E75'}

g = sns.pairplot(df_plot, hue='Overall_Impact', palette=palette,
                 plot_kws={'alpha': 0.5, 's': 20},
                 diag_kind='kde', corner=False)

g.figure.suptitle('Pairplot Fitur Numerik berdasarkan Kelas Overall_Impact',
                  y=1.02, fontsize=14, fontweight='bold')
plt.tight_layout()
plt.show()

print('''
📊 Interpretasi Pairplot:
  • Mental_Health_Score vs Avg_Daily_Usage_Hours:
    Kelas Positive cenderung memiliki skor kesehatan mental lebih tinggi
    dengan jam penggunaan yang lebih moderat.

  • Sleep_Hours_Per_Night:
    Kelas Negative cenderung memiliki jam tidur lebih sedikit,
    mengindikasikan media sosial berkorelasi dengan kurang tidur.

  • Mental_Health_Score:
    Distribusi KDE pada diagonal menunjukkan perbedaan distribusi yang
    jelas antar kelas — fitur ini berpotensi menjadi pembeda yang kuat.

  • Age & Avg_Daily_Usage_Hours tidak menunjukkan pemisahan kelas yang jelas,
    artinya pengaruhnya terhadap label mungkin kecil.
''')

smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_scaled, y)

print('=== Distribusi Kelas Sebelum SMOTE ===')
before = pd.Series(y.values).value_counts().sort_index()
for k, v in before.items():
    label = {0: 'Negative', 1: 'Neutral', 2: 'Positive'}[k]
    print(f'  {label}: {v}')

print('\n=== Distribusi Kelas Setelah SMOTE ===')
after = pd.Series(y_resampled).value_counts().sort_index()
for k, v in after.items():
    label = {0: 'Negative', 1: 'Neutral', 2: 'Positive'}[k]
    print(f'  {label}: {v}')

print(f'\nTotal data sebelum SMOTE: {len(y)}')
print(f'Total data setelah SMOTE: {len(y_resampled)}')

# Visualisasi perbandingan
fig, axes = plt.subplots(1, 2, figsize=(12, 4))
labels = ['Negative', 'Neutral', 'Positive']
colors = ['#E24B4A', '#888780', '#1D9E75']

axes[0].bar(labels, before.values, color=colors, edgecolor='white')
axes[0].set_title('Sebelum SMOTE', fontweight='bold')
axes[0].set_ylabel('Jumlah Data')
for i, v in enumerate(before.values):
    axes[0].text(i, v + 10, str(v), ha='center', fontweight='bold')

axes[1].bar(labels, after.values, color=colors, edgecolor='white')
axes[1].set_title('Setelah SMOTE', fontweight='bold')
axes[1].set_ylabel('Jumlah Data')
for i, v in enumerate(after.values):
    axes[1].text(i, v + 10, str(v), ha='center', fontweight='bold')

plt.suptitle('Perbandingan Distribusi Kelas Sebelum & Setelah SMOTE',
             fontsize=13, fontweight='bold')
plt.tight_layout()
plt.show()

X_train, X_test, y_train, y_test = train_test_split(
    X_resampled, y_resampled,
    test_size=0.2,
    random_state=42,
    stratify=y_resampled  # Pastikan distribusi kelas seimbang
)

print('=== Hasil Split Dataset ===')
print(f'Total data (setelah SMOTE) : {len(X_resampled)}')
print(f'Data training (80%)        : {len(X_train)}')
print(f'Data testing  (20%)        : {len(X_test)}')

print('\n=== Distribusi Kelas pada Training Set ===')
train_dist = pd.Series(y_train).value_counts().sort_index()
for k, v in train_dist.items():
    label = {0: 'Negative', 1: 'Neutral', 2: 'Positive'}[k]
    print(f'  {label}: {v}')

print('\n=== Distribusi Kelas pada Testing Set ===')
test_dist = pd.Series(y_test).value_counts().sort_index()
for k, v in test_dist.items():
    label = {0: 'Negative', 1: 'Neutral', 2: 'Positive'}[k]
    print(f'  {label}: {v}')

k_range = range(1, 31)
cv_scores = []
train_scores = []

for k in k_range:
    knn = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
    # 5-fold cross validation pada training set
    cv = cross_val_score(knn, X_train, y_train, cv=5, scoring='accuracy')
    cv_scores.append(cv.mean())
    knn.fit(X_train, y_train)
    train_scores.append(knn.score(X_train, y_train))

best_k = k_range[np.argmax(cv_scores)]
best_cv_score = max(cv_scores)

print(f'✅ K Optimal: {best_k}')
print(f'   Akurasi CV terbaik: {best_cv_score:.4f} ({best_cv_score*100:.2f}%)')

# Plot
plt.figure(figsize=(12, 5))
plt.plot(k_range, cv_scores, 'o-', color='#185FA5', label='CV Accuracy (5-Fold)', linewidth=2)
plt.plot(k_range, train_scores, 's--', color='#D85A30', label='Training Accuracy', linewidth=2, alpha=0.7)
plt.axvline(x=best_k, color='#1D9E75', linestyle='--', linewidth=2,
            label=f'K optimal = {best_k}')
plt.axhline(y=best_cv_score, color='#1D9E75', linestyle=':', alpha=0.5)
plt.scatter([best_k], [best_cv_score], s=150, zorder=5, color='#1D9E75',
            edgecolors='black', linewidth=1.5)
plt.xlabel('Nilai K', fontsize=12)
plt.ylabel('Akurasi', fontsize=12)
plt.title('Pencarian K Optimal — KNN', fontsize=14, fontweight='bold')
plt.legend(fontsize=11)
plt.xticks(k_range)
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

knn_best = KNeighborsClassifier(n_neighbors=best_k, metric='euclidean')
knn_best.fit(X_train, y_train)
y_pred = knn_best.predict(X_test)

print(f'Model KNN dilatih dengan K={best_k}')
print(f'Akurasi pada data testing: {accuracy_score(y_test, y_pred)*100:.2f}%')

class_names = ['Negative', 'Neutral', 'Positive']
cm = confusion_matrix(y_test, y_pred)

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Raw counts
disp1 = ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=class_names)
disp1.plot(ax=axes[0], cmap='Blues', colorbar=False)
axes[0].set_title('Confusion Matrix (Jumlah)', fontweight='bold')

# Normalized
cm_norm = confusion_matrix(y_test, y_pred, normalize='true')
disp2 = ConfusionMatrixDisplay(confusion_matrix=cm_norm, display_labels=class_names)
disp2.plot(ax=axes[1], cmap='Blues', colorbar=False)
axes[1].set_title('Confusion Matrix (Normalisasi)', fontweight='bold')

plt.suptitle(f'Confusion Matrix — KNN (K={best_k})', fontsize=14, fontweight='bold', y=1.02)
plt.tight_layout()
plt.show()

print('=== Confusion Matrix (Raw) ===')
cm_df = pd.DataFrame(cm, index=class_names, columns=class_names)
cm_df.index.name = 'Aktual'
cm_df.columns.name = 'Prediksi'
print(cm_df)

print(f'''
📊 Interpretasi Confusion Matrix:
  • Diagonal utama (kiri atas → kanan bawah) = prediksi BENAR
  • Di luar diagonal = prediksi SALAH (misklasifikasi)

  • Negative  → Benar: {cm[0][0]}, Salah sebagai Neutral: {cm[0][1]}, Salah sebagai Positive: {cm[0][2]}
  • Neutral   → Benar: {cm[1][1]}, Salah sebagai Negative: {cm[1][0]}, Salah sebagai Positive: {cm[1][2]}
  • Positive  → Benar: {cm[2][2]}, Salah sebagai Negative: {cm[2][0]}, Salah sebagai Neutral: {cm[2][1]}
''')

acc = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred, target_names=class_names, output_dict=True)

print('=' * 55)
print(f'          CLASSIFICATION REPORT — KNN (K={best_k})')
print('=' * 55)
print(classification_report(y_test, y_pred, target_names=class_names))

# Visualisasi metrik per kelas
metrics = ['precision', 'recall', 'f1-score']
x = np.arange(len(class_names))
width = 0.25
colors_metric = ['#185FA5', '#1D9E75', '#D85A30']

fig, ax = plt.subplots(figsize=(10, 5))
for i, (metric, color) in enumerate(zip(metrics, colors_metric)):
    vals = [report[cls][metric] for cls in class_names]
    bars = ax.bar(x + i * width, vals, width, label=metric.capitalize(),
                  color=color, edgecolor='white', linewidth=1)
    for bar, val in zip(bars, vals):
        ax.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.01,
                f'{val:.2f}', ha='center', va='bottom', fontsize=9, fontweight='bold')

ax.axhline(y=acc, color='#7F77DD', linestyle='--', linewidth=2,
           label=f'Accuracy = {acc:.4f}')
ax.set_xlabel('Kelas', fontsize=12)
ax.set_ylabel('Skor', fontsize=12)
ax.set_title(f'Metrik Evaluasi per Kelas — KNN (K={best_k})', fontsize=13, fontweight='bold')
ax.set_xticks(x + width)
ax.set_xticklabels(class_names)
ax.set_ylim(0, 1.15)
ax.legend(fontsize=11)
ax.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()

print('\n=== Ringkasan Akhir ===')
print(f"  Accuracy  : {acc:.4f} ({acc*100:.2f}%)")
print(f"  Precision : {report['weighted avg']['precision']:.4f}")
print(f"  Recall    : {report['weighted avg']['recall']:.4f}")
print(f"  F1-Score  : {report['weighted avg']['f1-score']:.4f}")

ratios = [0.3, 0.2, 0.1]
ratio_labels = ['70:30', '80:20', '90:10']
results = []

for ratio, label in zip(ratios, ratio_labels):
    Xtr, Xte, ytr, yte = train_test_split(
        X_resampled, y_resampled, test_size=ratio,
        random_state=42, stratify=y_resampled
    )
    knn_tmp = KNeighborsClassifier(n_neighbors=best_k, metric='euclidean')
    knn_tmp.fit(Xtr, ytr)
    ypred_tmp = knn_tmp.predict(Xte)
    rep = classification_report(yte, ypred_tmp, output_dict=True)
    results.append({
        'Rasio': label,
        'Train': len(Xtr),
        'Test': len(Xte),
        'Accuracy': accuracy_score(yte, ypred_tmp),
        'Precision': rep['weighted avg']['precision'],
        'Recall': rep['weighted avg']['recall'],
        'F1-Score': rep['weighted avg']['f1-score']
    })

results_df = pd.DataFrame(results)
print('=== Perbandingan Metrik Berdasarkan Rasio Split ===')
print(results_df.set_index('Rasio').round(4).to_string())

fig, ax = plt.subplots(figsize=(10, 5))
metrics_cmp = ['Accuracy', 'Precision', 'Recall', 'F1-Score']
x_cmp = np.arange(len(ratio_labels))
w = 0.2
colors_cmp = ['#185FA5', '#1D9E75', '#D85A30', '#7F77DD']
for i, (m, c) in enumerate(zip(metrics_cmp, colors_cmp)):
    ax.bar(x_cmp + i*w, results_df[m], w, label=m, color=c, edgecolor='white')
ax.set_xticks(x_cmp + w*1.5)
ax.set_xticklabels(ratio_labels)
ax.set_xlabel('Rasio Train:Test', fontsize=12)
ax.set_ylabel('Skor', fontsize=12)
ax.set_ylim(0, 1.15)
ax.set_title('Perbandingan Performa KNN pada Berbagai Rasio Split', fontsize=13, fontweight='bold')
ax.legend(fontsize=10)
ax.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()`
